src/rdkafka_partition.c

```diff
@@ -609,26 +609,23 @@ void rd_kafka_toppar_desired_del (rd_kafka_toppar_t *rktp) {
  * Append message at tail of 'rktp' message queue.
  */
 void rd_kafka_toppar_enq_msg (rd_kafka_toppar_t *rktp, rd_kafka_msg_t *rkm) {
+	int wakeup_fd, queue_len;
+	rd_kafka_t *rk;
 
 	rd_kafka_toppar_lock(rktp);
+	wakeup_fd = rktp->rktp_msgq_wakeup_fd;
+	rk = rktp->rktp_rkt->rkt_rk;
 	rd_kafka_msgq_enq(&rktp->rktp_msgq, rkm);
+	queue_len = rd_kafka_msgq_len(&rktp->rktp_msgq);
+        rd_kafka_toppar_unlock(rktp);
 #ifndef _MSC_VER
-        if (rktp->rktp_msgq_wakeup_fd != -1 &&
-            rd_kafka_msgq_len(&rktp->rktp_msgq) == 1) {
+        if (wakeup_fd >= 0 && queue_len == 1) {
```

Comment:
This is not completely without side-effects:

Having the queue len + write outside of holding the lock risks two (or more) threads calling write() simultaneously which will cause additional wake-ups in the broker thread and more context switches in these enq threads, all resulting in increased CPU usage.

Reply:
I don't think so?  The key thing is that the queue length is read while we hold the lock.  That means if two threads read a queue length of 1 in the time between when those two threads held the lock another thread must have dequeued messages to set the queue length to 0.  Another way to think about this is that if you put the unlock after the if the additional operations that take place with the lock held  only depend on values local to the executing thread.